geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/client/FilterAggregatingClient.java

```diff
@@ -0,0 +1,139 @@
+
+package org.locationtech.geomesa.hbase.client;
+
+import com.google.protobuf.RpcCallback;
+import com.google.protobuf.RpcController;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.client.coprocessor.Batch.*;
+import org.apache.hadoop.hbase.ipc.BlockingRpcCallback;
+import org.locationtech.geomesa.hbase.proto.KryoLazyDensityProto.*;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map.*;
+
+public class FilterAggregatingClient {
+
+    /**
+     * An RpcController implementation for use here in this endpoint.
+     */
+    static class FilterAggregatingClientRpcController implements RpcController {
+        private String errorText;
+        private boolean cancelled = false;
+        private boolean failed = false;
+
+        @Override
+        public String errorText() {
+            return this.errorText;
+        }
+
+        @Override
+        public boolean failed() {
+            return this.failed;
+        }
+
+        @Override
+        public boolean isCanceled() {
+            return this.cancelled;
+        }
+
+        @Override
+        public void notifyOnCancel(RpcCallback<Object> arg0) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void reset() {
+            this.errorText = null;
+            this.cancelled = false;
+            this.failed = false;
+        }
+
+        @Override
+        public void setFailed(String errorText) {
+            this.failed = true;
+            this.errorText = errorText;
+        }
+
+        @Override
+        public void startCancel() {
+            this.cancelled = true;
+        }
+
+    }
+
+    /**
+     * It gives the combined result of pairs received from different region servers value of a column for a given column family for the
+     * given range. In case qualifier is null, a min of all values for the given
+     * family is returned.
+     *
+     * @param table
+     * @return HashMap result;
+     * @throws Throwable
+     */
+    public List<Pair> kryoLazyDensityFilter(final Table table) throws Throwable {
+
+        final DensityRequest requestArg = DensityRequest.newBuilder().build();
+        class KryoLazyDensityFilterCallBack implements Callback<List<Pair>> {
+
+            private HashMap<String, Double> finalResult = new HashMap<String, Double>();
+
+            List<Pair> getResult() {
+                List<Pair> list = new ArrayList<Pair>();
+                for (Entry e : finalResult.entrySet()) {
+                    Pair pair = Pair.newBuilder().setKey((String) e.getKey()).setValue((Double) e.getValue()).build();
+                    list.add(pair);
+                }
+
+                return list;
+            }
+
+            @Override
+            public synchronized void update(byte[] region, byte[] row, List<Pair> result) {
```

Comment:
No. Update is called when everything is being run and when updates are complete, only then is getResult called.